### PR TITLE
Materialized Views for `sqlalchemy-redshift`

### PR DESCRIPTION
**Pull request by ewengillies**
I will be creating a DDLElement that can support Redshift's [CREATE MATERIALIZED VIEW](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html) with table attributes, noteably sortkeys and distkeys/styles.

## Todos
 - [x] Factored out and tested `table_attributes` parser
 - [x] Write the CreateMaterializedView DDLElement
 - [x] Write the DDL compiler
 - [x] Test on redshift db

## Default Todos
- [x] MIT compatible
- [x] Unit Tests
- [x] Documentation
- [x] Updated CHANGES.rst

Notes:
 * I made the `table_attributes` parser less permissive to match expected outcomes from Redshift.  I'm happy to relax this and delegate the error messages to Redshift, but it could be nice to catch them early.  Up to the maintainers.
 * It looked like the same parsing code wanted to allow `Columns` in additions to strings (and iterables thereof) to be allowed to specify keys.  I added the checked needed in if this is the case.
 * I took the liberty of making a new module `ddl`, since the `Create` and `Drop` queries are normally treated somewhat differently than `Executable` or `ClauseElement`.  This was just an observation of mind, i'm happy to move it all into `commands`.
 * Unfortunately `REFRESH MATERIALIZED VIEWS` **is** a command.  So my contribution is a little split between these two modules, perhaps awkwardly.  Happy to take advice here.
 * Finally, I made some changes to the `doc` folder to reflect the new module.  I haven't used Sphinx so much, but it all seemed pretty self explanatory.  Even so, I'm not sure how to generate the docs to check locally.  Please advise, thanks!

### CODE REVIEW

**Comment by ewengillies** [comment]
@jklukas , here is the MR we discussed on #201 .  Awaiting your review, no rush tho!  Thanks :)

File: CHANGES.rst [hunk 1]
```diff
@@ -1,3 +1,9 @@
+0.7.9 (unreleased)
+------------------
+
+- Added support for materialized views
+  (`AWS Docs <https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-overview.html`_)
```
Comment (jklukas) [inline CHANGES.rst:5]: This should be added under the existing 0.7.8 section, since that release has not yet been finalized and this change will go out with 0.7.8.

File: sqlalchemy_redshift/ddl.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
+import sqlalchemy as sa
```
Comment (jklukas) [inline sqlalchemy_redshift/ddl.py:1]: Can you tell me a bit more about the decision to break out a `ddl.py`? Does the PostgreSQL dialect or other dialects do this?

File: CHANGES.rst (hunk 1 above)
Comment (ewengillies) [reply CHANGES.rst:5]: okay will do.

File: sqlalchemy_redshift/ddl.py (hunk 2 above)
Comment (ewengillies) [reply sqlalchemy_redshift/ddl.py:1]: My rough understanding is that `DDL` ~= `Create/Drop` clauses.  In this spirit, I think they are different from the `_ExecutableClauses` found in `commands.py`.  This is my main reason for splitting; its pretty pedantic.  I can definitely see the reasoning behind keeping it all in one `commands.py` file for simplicity.  

As to leave no stone unturned, I've looked closer and found that `ddl.py` has its own module under `sqlalchemy/sql/ddl.py`, whereas query clauses seem to be mostly in `sqlalchemy/sql/selectable.py`.

Once we go into the dialect tho, `sqlalchemy/dialects/postgresql`, its the `base.py` where most of the dropping and creating happens (i.e. contains the DDL modifications for PostgreSQL).  Under  `sqlalchemy/dialects/postgresql`, there seem to be special modules for some of PostgreSQL's special types (json, array, etc.).

In both cases, I fail to find the direct analogy of Redshift's "commands" concept.  However, I think that `commands` is apt for describing `COPY`, `UNLOAD`, and the other existing function in `commands.py`.  I'm pretty new to this level of `sqlalchemy` tho, so I could just be missing something.

In summary, we can either:
 * separate out the concept of Redshift-specific "creation" commands to `ddl` vs.
 * keep it all in `commands`, just to keep it simple.

happy to take your lead, I don't mind spending the time for option 2 if you think its best.

File: sqlalchemy_redshift/commands.py [hunk 3]
```diff
@@ -931,3 +931,45 @@ def visit_create_library_command(element, compiler, **kw):
                          or_replace='OR REPLACE' if element.replace else '',
                          region='REGION :region' if element.region else '')
     return compiler.process(sa.text(query).bindparams(*bindparams), **kw)
+
+
+class RefereshMaterializedView(_ExecutableClause):
```
Comment (jklukas) [inline sqlalchemy_redshift/commands.py:936]:
```suggestion
class RefreshMaterializedView(_ExecutableClause):
```

File: CHANGES.rst (hunk 1 above)
Comment (ewengillies) [reply CHANGES.rst]: done.

File: sqlalchemy_redshift/ddl.py (hunk 2 above)
Comment (jklukas) [reply sqlalchemy_redshift/ddl.py:1]: Thanks for the research here. I suppose it comes down to what makes most sense for the user when importing. Should we prefer:

     from sqlalchemy_redshift.ddl import CreateMaterializedView

Or:

     from sqlalchemy_redshift import CreateMaterializedView

Having it under `ddl` seems reasonable. I agree with you that DDL statements seem different from commands, so separating them is a good idea. And I think I also like the idea that we keep "extras" like these outside of the main `dialect.py` filed.

Comment (ewengillies) [reply sqlalchemy_redshift/ddl.py:1]: well right now, I cannot import `CopyCommand` from `sqlalchemy_redshift`, but I can import it from `sqlalchemy_redshift.dialect` or `sqlalchemy_redshift.commands`.

If `CreateMaterializedView` lives in `ddl.py`, then analogously it can be imported from `sqlalchemy_redshift.dialect` or `sqlalchemy_redshift.ddl`.  This is because I followed the example of the package, where `dialect.__all__` has all commands and types.  

This follows the pattern in `sqlalchemy/dialects/postgresql/__init__.py`, where `__all__` contains elements like `BIGINT` and `CreateEnumType`.  The import pattern in SQLAlchemy seems to be like:
```python
from sqlalchemy.dialects.postgresql import X
```

perhaps then, something like
```python
from sqlalchemy_redshift.dialect import X
```
is not too different.

My proposal is then: 
 * we can point the users to `from sqlalchemy_redshift.dialect import X`.
 * If this package grows, and we want to factor out a `base.py`, etc, `dialect.py` can become a (depreciatable) high-level namespace that contains all the dialect elements, much like `sqlalchemy/dialects/postgresql/__init__.py`.

Comment (jklukas) [reply sqlalchemy_redshift/ddl.py:1]: I appreciate the thought you've put into this. I leave it up to you to decide what the "preferred" import should look like (ddl vs. dialect), and make sure that the docstring example and the tests import it using the preferred style.

Let me know when you consider this done, and I'll run a final integration test and merge.

Comment (ewengillies) [reply sqlalchemy_redshift/ddl.py:1]: okay, I'm done, I changed any `commands` or `ddl` imports in the tests to the `dialects` namespace.  I likewise updated the docs that are in the docstrings to reflect this change as well.

**Comment by jklukas** [comment]
I just posted a new release 0.7.9 which includes this change.
